umbraco/Umbraco-CMS
# V14: Start using user group permission service

## Details
- As part of implementing authorization policies for the Management API, we will start introducing dedicated `XAuthorizationStatus` and `IXPermissionService`;
  - For user groups, we have already introduced `IUserGroupAuthorizationService` for handling some of the logic in v14, so this PR is:
    - Moving methods from `IUserGroupAuthorizationService` to `IUserGroupPermissionService` & removing it;
    - Refactoring to use `UserGroupAuthorizationStatus` for all authorization things (instead of `UserGroupOperationStatus`);
    - Refactor to return only `UserGroupAuthorizationStatus`, without the `Attempt`, as it is not used.
- We are also mapping `UserGroupAuthorizationStatus` to `UserGroupOperationStatus` as the Create and Update User group controllers expect `UserGroupOperationStatus` which can cover many other things than validation results.

## Test
- Controllers to test:
  - Create User Group: `POST /umbraco/management/api/v1/user-group`;
  - Update User Group: `PUT /umbraco/management/api/v1/user-group/{id}`.
- What to test _(have this in mind when reading "How to test"):_
  - Have a user part of a user group;
  - Modify the user group to test different scenarios;
  - Authorize against the Management API to get a new token every time the user group permissions change;
  - Modify the payload of the requests.
- How to test - ❕ I ran the **Umbraco.Web.UI** for doing the setup and modifications of the user group permissions through the old backoffice _(remember to run `npm run build` on **Umbraco.Web.UI.Login** to make the auth work):_
  1. Comment out the authorization policy on `UserGroupControllerBase`- _[this line](https://github.com/umbraco/Umbraco-CMS/blob/v14/dev/src/Umbraco.Cms.Api.Management/Controllers/UserGroup/UserGroupsControllerBase.cs#L14)_, so you make sure that the policies are not interfering and the refactored logic still works (`IUserGroupPermissionService.AuthorizeCreateAsync()` & `IUserGroupPermissionService.AuthorizeUpdateAsync()`);
  2. Create a simple user group with **just** a few sections, **other** ❗  than **Content**, **Media** and **Users** ❗; 
  3. Create a user part of this newly created user group;
  4. Create a content node at root and a media node at root and keep in mind their ids for the "create" payload later (for `documentStartNodeId` and `mediaStartNodeId`);
  5. Authorize against the Management API with that user;
  6. Try to **create** a user group with the following payload:
```json
{
  "name": "UserGroupTest1",
  "sections": [
	"Umb.Section.Packages",
	"Umb.Section.Translation"
  ],
  "languages": [],
  "hasAccessToAllLanguages": true,
  "documentStartNodeId": "b4ad9e02-28b8-4f4c-a1b8-3302d6ae0762",
  "mediaStartNodeId": "6224aae8-8fc3-4172-809a-d7b2a86d15ca",
  "permissions": []
}
``` 

  - 
    7. Verify that:
      - Since your user doesn't have access to the Users section (the user group it is part of doesn't have Users in its allowed sections, as pointed out in step **ii.**);
        - you will get **401 Unauthorized** with the message "**Unauthorized user group access**".
      - Modify the initial user group of the user (_as admin through the backoffice_) to include the "**Users**" section;
      - And make sure that there is a section in the payload that the user doesn't have access to, before the next step (_ex: make sure that the Packages section is not listed in the allowed sections of the initial user group of the user_);
      - Repeat step **vi.**;
        - you will get **401 Unauthorized** with the message "**Unauthorized section access**".
      - Modify the initial user group of the user (_as admin through the backoffice_) to include the missing section;
      - Repeat step **vi.**;
        - you will get **401 Unauthorized** with the message "**Unauthorized content start node access**".
      - Modify the initial user group of the user (_as admin through the backoffice_) to have Content as an allowed section and add a "Content start node" as the root of the content tree;
      - Repeat step **vi.**;
        - you will get **401 Unauthorized** with the message "**Unauthorized media start node access**".
      - Modify the initial user group of the user (_as admin through the backoffice_) to have Media as an allowed section and add a "Media start node" as the root of the media tree;
      - Repeat step **vi.**;
        - It will return 201 - verify that the user group has been created by fetching it.
    8. The same things can be validated by modifying the user group permissions through the backoffice and then using such payload to the **update** a user group request;
      - There is a special check that is different to what we've already seen:
        - Create a different user group through the backoffice - the user you are authenticating against the Management API **shouldn't** be part of that group;
        - Get its id and execute PUT /umbraco/management/api/v1/user-group/{id};
          - you will get **401 Unauthorized** with the message "**Unauthorized user group access**".
